### ISSUE TITLE
Fix the attribute order problem

### ISSUE THREAD

**Issue by chrisdone** [chrisdone/lucid#110]
For #107 and #109 it brings to light that the instability of Data.HashMap is really annoying for writing tests.

Options:

* Just use Data.Map. Slower, but stable. Easy drop in replacement.
* Use [InsOrdHashMap.](https://hackage.haskell.org/package/insert-ordered-containers-0.2.5/docs/Data-HashMap-Strict-InsOrd.html) I haven't actually tested the performance of this. It's supposed to be "HashMap but ordered". They use it at Hasura, but I haven't personally benchmarked it. Easy drop in replacement, though.
* Use `[Attribute]` and `ordNub` and assume nobody will use the concatenation of attributes feature.
* Stick with HashMap, but throw a `sortBy (comparing fst)` just after `HM.toList`. Could be faster than `Data.Map`, but would need benching.

I don't really care about raw performance advantages anymore; the maintenance burden is wasting other people's time and by extension our own.

[This blog post](https://oleg.fi/gists/posts/2021-05-19-dont-default-to-hashmap.html) is a bit presumptuous (assumes that I didn't benchmark HashMap vs Map, which I did, and it is faster for many attributes) and talks out of its arse a bit about what is and isn't fast--as Haskellers are wont to do--but nonetheless, the lack of ordering is truly annoying.

**Comment by phadej** [comment]
I tried changing the `HashMap` to `Map`, and benchmark results changed as:

```
Benchmark                    bench.master  bench2.master    bench.map            bench2.map         
basic/ByteString             0.576e-5      0.575e-5 -0.14%   0.591e-5    +2.71%   0.589e-5    +2.26%
bigTable/ByteString          2.303e-3      2.303e-3 +0.03%   2.363e-3    +2.59%   2.328e-3    +1.07%
customAttribute/ByteString   0.615e-2      0.614e-2 -0.21%  19.768e-2 +3114.92%  18.424e-2 +2896.30%
deepTree/ByteString          0.862e-3      0.857e-3 -0.61%   0.865e-3    +0.34%   0.867e-3    +0.54%
manyAttributes/ByteString    1.141e-2      1.134e-2 -0.56%  87.991e-2 +7613.09%  88.448e-2 +7653.15%
wideTree/ByteString          2.078e-3      2.089e-3 +0.53%   2.398e-3   +15.41%   2.370e-3   +14.09%
wideTreeEscaping/ByteString  0.391e-3      0.392e-3 +0.19%   0.450e-3   +15.14%   0.445e-3   +13.75%
Geometric mean               0.940e-3      0.939e-3 -0.11%   3.016e-3  +220.63%   2.970e-3  +215.80%
```

I'm very surprised why `Map` is so slow on manyattributes benchmark, as its only 5000 elements, if I understand the benchmark right.

**Comment by phadej** [comment]
Hmm,

```haskell
manyAttributes :: [String]  -- ^ List of attribute values.
               -> Html ()      -- ^ Result.
manyAttributes as = img_ (map (id_ . T.pack) as)
```

Doesn't that create a single attribute (`id`), and rather concatenates the list of strings?

**Comment by chrisdone** [comment]
Yeah, probably a better name for this BlazeMarkup test would be "duplicateAttributes".

**Comment by phadej** [comment]
The description says 

>         "A single element with " >> toHtml (show (length manyAttributesData))
        " attributes."
        
---

With a small change

```diff
-manyAttributesData :: [String]
-manyAttributesData = wideTreeData
+manyAttributesData :: [(T.Text, T.Text)]
+manyAttributesData = zipWith
+    (\i x -> (T.pack ("attr" ++ show i), T.pack x))
+    [0 :: Int ..]
+    wideTreeData

 -- | Create an element with many attributes.
 --
-manyAttributes :: [String]  -- ^ List of attribute values.
+manyAttributes :: [(T.Text, T.Text)]  -- ^ List of attribute values.
                -> Html ()      -- ^ Result.
-manyAttributes as = img_ (map (id_ . T.pack) as)
+manyAttributes as = img_ (map (\(k, v) -> makeAttribute k v) as)
```

(packing in data already, as we don't need to benchmark `T.pack`)

The results are a lot closer:

```
Benchmark                    bench4.master  bench4.map         
basic/ByteString             0.584e-5        0.593e-5    +1.44%
bigTable/ByteString          2.202e-3        2.223e-3    +0.94%
customAttribute/ByteString   0.612e-2       19.552e-2 +3096.45%
deepTree/ByteString          0.867e-3        0.852e-3    -1.71%
manyAttributes/ByteString    0.438e-2        0.553e-2   +26.20%
wideTree/ByteString          2.064e-3        2.203e-3    +6.72%
wideTreeEscaping/ByteString  0.380e-3        0.404e-3    +6.19%
Geometric mean               0.813e-3        1.405e-3   +72.80%
```

customAttribute still exercises something in (small!) `Map`s which make them perform bad.

---

I'll make a PR adding the true *many attributes* benchmark and renaming & adjusting the description of current one (which as far as I can tell should be dominated by text concatenation, we should investigate why `Map` regresses there so badly).

**Comment by chrisdone** [comment]
Fixed in #113.
